src/chai/ArrayManager.cpp

```diff
@@ -137,16 +145,24 @@ void ArrayManager::move(PointerRecord* record, ExecutionSpace space)
     return;
   }
 
+  if (space == record->m_last_space) {
+    return;
+  }
+
 #if defined(CHAI_ENABLE_UM)
   if (record->m_last_space == UM) {
     return;
   }
 #endif
 
-  if (space == record->m_last_space) {
+#if defined(CHAI_ENABLE_PINNED)
+  if (record->m_last_space == PINNED) {
+    if (space == CPU) {
+      cudaDeviceSynchronize();
```

Comment:
As written, I believe this will cause a synchronize per array. Can we add some state to the resource manager that remembers it's already done a sync that gets reset when there is any move to a GPU space?